r2.9.4

ojw28 opened the pull request.

tonihei committed:
Fix manifest uri in SsDownloadHelper.

This is the same as in SsMediaSource.

PiperOrigin-RevId: 225001911

AquilesCanta committed:
Update the Cast framework dependency

PiperOrigin-RevId: 225812585

AquilesCanta committed:
Handle failure to get Cast context more gracefully

Issue:#4160
Issue:#4743
PiperOrigin-RevId: 225813243

tonihei committed:
Improve doc for setKeepContentOnPlayerReset.

This also applies when seeking or transitioning to unprepared media, which
isn't clear from the current documentation.

Issue:#5267
PiperOrigin-RevId: 226486685

andrewlewis committed:
Handle rectangular rotation projections in Matroska

See also https://github.com/Matroska-Org/matroska-specification/issues/269.

PiperOrigin-RevId: 226758584

ojw28 committed:
Remove stray word

PiperOrigin-RevId: 227500707

andrewlewis committed:
Remove AdsLoader listeners on releasing ImaAdsLoader

Issue: #4114
PiperOrigin-RevId: 227516509

ojw28 committed:
Move syncFileDescriptor to use an experimental method

PiperOrigin-RevId: 227520168

andrewlewis committed:
Fix replacement char check

PiperOrigin-RevId: 227646358

tonihei committed:
Use Handler instead of ExoPlayer messages in ConcatenatingMediaSource

ExoPlayer methods must not be called from any thread besides the specified
app thread. Therefore we shouldn't use them here. Using a regular Handler
instead is fully equivalent.

Issue:#5240
PiperOrigin-RevId: 227650489

AquilesCanta committed:
Fix bug when calculating EOF position in mp4 sniffing

PiperOrigin-RevId: 227668426

AquilesCanta committed:
Enable setOutputSurfaceWorkaround for Huawei P10 lite

Issue:#5312
PiperOrigin-RevId: 227673949

AquilesCanta committed:
Prevent IllegalStateException in Mp4 sniffing

If a negative value is read, sniffing should just fail.

PiperOrigin-RevId: 227689568

andrewlewis committed:
Parse frame rate from 'mdta' metadata

PiperOrigin-RevId: 227813461

andrewlewis committed:
Fix nullness issue

PiperOrigin-RevId: 227822937

andrewlewis committed:
Treat AVERROR_INVALIDDATA as non-fatal

Also configure the FFmpeg context to ignore errors as far as possible (this
appears to have an effect only for certain decoders).

Issue: #5293
PiperOrigin-RevId: 227851397

andrewlewis committed:
Update FakeAd overrides

These are part of published IMA SDK 3.10.2.

PiperOrigin-RevId: 227861713

AquilesCanta committed:
Increase search size in mp4 sniffing once moov has been found

Issue:#5320
PiperOrigin-RevId: 228142567

AquilesCanta committed:
Expand check for muxed audio media tags to include uris that match variants

Issue:#5313
PiperOrigin-RevId: 228155222

ojw28 committed:
Doc fix.

PiperOrigin-RevId: 228296962

tonihei committed:
Update README and dev guide with Java 8 config for Kotlin.

Setting the target conpatibility only seems to work for Java. Added the
equivalent Kotlin config options to the docs.

Issue:#5276
PiperOrigin-RevId: 228482496

tonihei committed:
Add missing call to timeline.getWindow.

The window object is used without being filled with data. This used to work
well for most cases as the same live stream is sending regular updates and the
first update is almost never used if it's not the first item in a playlist.

It causes problems when the first timeline update of a live stream is actually
used for playback (e.g. when the live stream is lazily prepared in a playlist
and played first).

PiperOrigin-RevId: 228530232

andrewlewis committed:
￼Blacklist OMX.SEC.mp3.dec for more devices

Issue: #4519
PiperOrigin-RevId: 229145790

tonihei committed:
Don't forget isSeekable in ExtractorMediaSource.

We currently forget whether a source is seekable at re-preparation. This was
implemented intentionally this way under the assumption that we really can't seek
until we have loaded the seek map again. However, seek operations are only
allowed after a media period is prepared. So there is no harm in remembering
whether a source is seekable.

This problem currently prevents reusing ClippingMediaSources with
ExtractorMediaSource and a non-zero start clip position.

Issue: #5351
PiperOrigin-RevId: 229169441

ojw28 committed:
Fix DRM protected SmoothStreaming with subtitles

Issue: #5378
PiperOrigin-RevId: 229261658

tonihei committed:
Add buffer flag for last sample to improve buffered position calculation.

The buffered position is currently based on the mimimum queued timestamp of
all AV tracks. If the tracks have unequal lengths, one track continues loading
without bounds as the "buffered position" will always stay at the shorter
track's duration.

This change adds an optional buffer flag to mark the last sample of the
stream. This is set in the Mp4Extractor only so far. ExtractorMediaSource
uses this flag to ignore AV streams in the buffered duration calculation if
they already finished loading.

Issue:#3670
PiperOrigin-RevId: 229359899

ojw28 committed:
Fix release branch

ojw28 committed:
Bump version for 2.9.4 release

PiperOrigin-RevId: 229364563

tonihei committed:
Fix typo.

PiperOrigin-RevId: 229365333

tonihei committed:
Fix bug where missing switch adaptation set causes multiple identical track groups.

When the extra adaptation set of a switch group isn't defined in the manifest, we
currently assume it's the first adaptation group. This either leads to wrong grouping
or duplicate track groups.

Such a case may easily happen if the manifest is filtered such that only one of the
switch adaptation sets will be present in the manifest.

PiperOrigin-RevId: 229365379

marcbaechinger committed:
allow developers to set the subText of the notifcation

Issue: #5344
PiperOrigin-RevId: 229527963

mr-thierry commented:
Is there anything blocking this release? I'm waiting for the fix for [#4114](https://github.com/google/ExoPlayer/issues/4114) to release my app..

ojw28 commented:
We have one or two things that we want to add still. It'll be soon though (probably early next week).

tonihei committed:
Add start position to MediaSource.createPeriod.

That's the same position set in MediaPeriod.prepare (where it may be removed
in the future).

Having the position at an earlier point is necessary to fix an
issue with lazy preparation in ConcatenatingMediaSource where the prepare
position was assumed to be known but MediaPeriod.prepare hasn't been called
yet.

Issue:#5350
PiperOrigin-RevId: 229756637

tonihei committed:
Add missing @Nullable to SimpleExoPlayer fields and methods.

Issue:#5402
PiperOrigin-RevId: 229758525

andrewlewis committed:
Upgrade to GVR SDK 1.190.0

Change the dependency to the new monolithic GVR SDK target.

PiperOrigin-RevId: 229931549

tonihei committed:
Add max video size workaround for Amlogic decoder.

The Amlogic awesome decoder reduces the video size of interlaced videos by half
if the internal configuration isn't force reset with new maximum input size
values. The product of these new values must exceed 1920x1088 to force the
reset.

Issue:#5003
PiperOrigin-RevId: 230206675

andrewlewis committed:
Fix GVR dependency

PiperOrigin-RevId: 230213842

tonihei committed:
Use loading period event time for fatal load errors.

ExoPlaybackExceptions of type SOURCE are always associated with the loading
period and thus we can use the event time for the loading period in
onPlayerError. Renderer and unexpected exceptions are still associated with the
currently playing period.

Issue:#5407
PiperOrigin-RevId: 230216253

ojw28 committed:
Fix release notes
